presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java

```diff
@@ -589,6 +591,8 @@ else if (taskState == TaskState.FINISHED) {
                 finishedTasks.add(taskStatus.getTaskId());
             }
 
+            // The finishedTasks.add(taskStatus.getTaskId()) must happen before the getState() (see schedulingComplete)
+            stageExecutionState = getState();
```

Comment:
If we don't refresh the state here, will the following `if` block still transition the state to finished?

Reply:
No. We need to change for the state after we update the `finishedTasks` array.

Let's assume we check the status before we update the `finishedTasks`:

e.g.:

```
updateTaskStatus:

- check state
- update finished tasks
- transition to FINISHED if the state is SCHEDULED and all tasks are FINISHED

transitionToScheduled:

schedulingComplete:

- transition to SCHEDULED
- transition to FINISHED when all tasks are FINISHED

```

next execution is possible:

```
- updateTaskStatus: -> check state (not SCHEDULED)
- schedulingComplete: -> transition to SCHEDULED
- schedulingComplete: -> check if all tasks are finished -> some still not finished -> not transition
- updateTaskStatus: -> update finished tasks
- updateTaskStatus: -> state is not `SCHEDULED` -> not transition

```

The end stage state after this missfortunate execution is not `FINISHED`, but `SCHEDULED`.

Now, let's change the order of the `update finished tasks` and the `check state`

```
updateTaskStatus:

- update finished tasks
- check state
- transition to FINISHED if the state is SCHEDULED and all tasks are FINISHED

transitionToScheduled:

schedulingComplete:

- transition to SCHEDULED
- transition to FINISHED when all tasks are FINISHED

```

With this ordering the execution that least to stage not being transitioned to FINISHED is not possible.

Reply:
I see. So if my understanding is correct, it means that it will take one more update cycle to finally transition the state to FINISHED for the unfortunate execution sequence. Thanks for clarifying.

Reply:
> it means that it will take one more update cycle to finally transition the state to FINISHED

Why? I think it shouldn't.

So conceptually there are two places that can "finish" the stage.

1. Transition to scheduled (if all the tasks are finished)
2. Update task statues (if the stage is in SCHEDULED state, and this is the update for the last task that is finished)

Whatever becomes true last - will transition the stage to FINISHED.

Reply:
Ah I should explain my understanding better. I mean if we don't check state again after update finished tasks, it will take another update if unfortunate sequence happens because it will transition to SCHEDULED while ideally it should directly go to FINISHED. 

So this change would save one update.

Reply:
One extra update is not a problem. If I understand the code correctly, if the last task reports "FINISHED", there may be no more updates. So the stage is at risk of being in the SCHEDULED state forever.

Reply:
Oh that's the key catch!

Reply:
Yeah. Before this property was enforced by the `synchronized`.

Reply:
@arhimondr , @shixuan-fan : I am wondering would it help to put line 595-603 to be in critical section? (i.e. synchronized on "this") -- this seems easier to reason?

Reply:
We can do this as a protective measure. But there should be no reason to do so. The execution is already ordered. And as long as the state is transitioned to `FINISHED` I'm not sure if we need stronger ordering requirements.

Reply:
@arhimondr 
> But there should be no reason to do so. 

Makes the system easier to reason can also be a "reason" 😃

Reply:
Yeah. I agree with that. I'm just not sure that having a synchronized here will introduce more clarity either. I spent a lot of time analyzing this code before I remove the existing synchronized in the first place. Next time when somebody will try to understand that code they might get confused again "why the synchronized is needed".